### PR TITLE
Start removing std::regex

### PR DESCRIPTION
**Pull request by strega-nil**

### CODE REVIEW

File: include/vcpkg/base/util.h [hunk 1]
```diff
@@ -130,12 +130,57 @@ namespace vcpkg::Util
...
+    // so find_nth({1, 2, 1, 3, 1, 4}, 1, 2)
+    // returns the 1 before the 4
+    template <class R, class V>
+    auto find_nth(R&& r, const V& v, size_t n)
+    {
+        using std::begin;
+        using std::end;
+
+        auto last = end(r);
+        auto current = std::find(begin(r), last, v);
```
Comment (BillyONeal) [inline include/vcpkg/base/util.h:154]: I doubt you care but the STL name for this would be "first" rather than "current". ([first,last) is for ranges, [begin(), end()) for containers over there).

File: include/vcpkg/base/util.h [hunk 2]
```diff
@@ -130,12 +130,57 @@ namespace vcpkg::Util
...
+        }
+
+        return current;
+    }
+
+    // 0th is the last occurence
+    // so find_nth({1, 2, 1, 3, 1, 4}, 1, 2)
+    // returns the 1 before the 2
+    template <class R, class V>
+    auto find_nth_from_last(R&& r, const V& v, size_t n)
```
Comment (BillyONeal) [inline include/vcpkg/base/util.h:168]:
```suggestion
    auto find_nth_from_last(R& r, const V& v, size_t n)
```

?

You never want the caller to move their range into this since you're returning an iterator into that.

Comment (BillyONeal) [reply include/vcpkg/base/util.h:168]: Also, I think you need to better consider what happens when the Nth occurrence isn't found; I think you really want that to return end but I think this returns begin instead?

File: include/vcpkg/base/util.h [hunk 3]
```diff
@@ -130,12 +130,57 @@ namespace vcpkg::Util
...
+
+        auto last = rend(r);
+        auto current = std::find(rbegin(r), last, v);
+
+        for (size_t i = 0; i < n && current != last; ++i)
+        {
+            current = std::find(current + 1, last, v);
+        }
+
+        return current.base() - 1;
```
Comment (BillyONeal) [inline include/vcpkg/base/util.h:181]: I thought you didn't want a reverse_iterator solution? How about something like this:

```
template<class BidIt, class NeedleTy>
BidIt find_nth_from_last(BidIt first, const BidIt last, const NeedleTy& needle, size_t n) {
    if (first != last) {
        BidIt next = last;
        do {
            --next;
            if (needle == *next) {
                if (n == 0) {
                    return next;
                }

                --n;
            }
        } while (first != next);
    }

    return last;
}

template <class R, class NeedleTy>
auto find_nth_from_last(R& haystack, const NeedleTy& needle, size_t n) {
    using std::begin;
    using std::end;
    return find_nth_from_last(begin(haystack), end(haystack), needle, n);
}
```

https://gcc.godbolt.org/z/Yc9PT1f5c

File: include/vcpkg/postbuildlint.buildtype.h [hunk 4]
```diff
@@ -5,61 +5,32 @@
...
 
-        constexpr operator BackingEnum() const { return backing_enum; }
+        bool has_crt_linker_option(StringView sv) const;
+        StringLiteral to_string() const;
 
-        const Build::ConfigurationType& config() const;
-        const Build::LinkageType& linkage() const;
-        const std::regex& crt_regex() const;
-        const std::string& to_string() const;
+        static BuildType VALUES[4];
```
Comment (BillyONeal) [inline include/vcpkg/postbuildlint.buildtype.h:23]: I think it would be better to have this static in the .cpp.

File: src/vcpkg-test/util-tests.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#include <catch2/catch.hpp>
+
+#include <vcpkg/base/util.h>
+
+namespace Util = vcpkg::Util;
+
+TEST_CASE ("find_nth", "[util]")
```
Comment (BillyONeal) [inline src/vcpkg-test/util-tests.cpp:7]: Please add a test for "empty" to protect against off-by-one errors to these. (I think I remember an unprotected base()-1 in the product code?)

File: src/vcpkg/archives.cpp [hunk 6]
```diff
@@ -16,22 +16,34 @@ namespace
...
-        std::smatch match;
-        const bool has_match = std::regex_match(stem, match, std::regex{R"###(^(.+)\.(\d+\.\d+\.\d+)$)###"});
-        Checks::check_exit(
-            VCPKG_LINE_INFO, has_match, "Could not deduce nuget id and version from filename: %s", archive);
+        auto dot_after_name = Util::find_nth_from_last(stem, '.', 2);
+
+        auto is_digit_or_dot = [](char ch) {
+            return ch == '.' || Parse::ParserBase::is_ascii_digit(ch);
+        };
+        if (dot_after_name == stem.end() || !std::for_each(dot_after_name, stem.end(), is_digit_or_dot))
```
Comment (ras0219-msft) [inline src/vcpkg/archives.cpp:39]: I don't understand why we're testing the return value of `for_each()` here -- did you mean `!std::all_of()`?

File: src/vcpkg/archives.cpp [hunk 7]
```diff
@@ -16,22 +16,34 @@ namespace
...
+        {
+            Checks::msg_exit_with_message(
+                VCPKG_LINE_INFO, msgCouldNotDeduceNugetIdAndVersion, msg::path = archive);
+        }
 
-        const std::string nugetid = match[1];
-        const std::string version = match[2];
+        auto nugetid = StringView{stem.begin(), dot_after_name};
+        auto version = StringView{dot_after_name + 1, stem.end()};
```
Comment (ras0219-msft) [inline src/vcpkg/archives.cpp:46]:
```suggestion
        StringView nugetid{stem.begin(), dot_after_name};
        StringView version{dot_after_name + 1, stem.end()};
```

File: src/vcpkg/archives.cpp (hunk 6 above)
Comment (strega-nil) [reply src/vcpkg/archives.cpp:39]: ... yes

File: src/vcpkg/archives.cpp (hunk 7 above)
Comment (strega-nil) [reply src/vcpkg/archives.cpp:46]: I really don't like using direct initialization, and I avoid it everywhere unless it's necessary (which it never is now, because we require C++17)

File: include/vcpkg/postbuildlint.buildtype.h (hunk 4 above)
Comment (strega-nil) [reply include/vcpkg/postbuildlint.buildtype.h:23]: I don't understand this - what do you mean? It is in the .cpp.

File: include/vcpkg/base/util.h (hunk 3 above)
Comment (strega-nil) [reply include/vcpkg/base/util.h:181]: It was more about less thinky when using the thing, not necessarily removing all uses of reverse iterators in the implementation code.

File: src/vcpkg/binarycaching.cpp [hunk 8]
```diff
@@ -2168,29 +2156,119 @@ ExpectedS<std::vector<std::unique_ptr<IBinaryProvider>>> vcpkg::create_binary_pr
...
-    if (std::regex_match(version.cbegin(), version.cend(), sm, date_matcher))
+    // /(\d\d\d\d)-(\d\d)-(\d\d).*/
+    bool try_parse_reformat_date_version(ReformatParsedVersion& out, StringView version)
+    {
+        using P = vcpkg::Parse::ParserBase;
+        // a b c d - e f - g h <end>
+        // 0 1 2 3 4 5 6 7 8 9 10
+        if (version.size() < 10) return false;
+        auto first = version.begin();
+        if (!P::is_ascii_digit(*first++)) return false;
```
Comment (BillyONeal) [inline src/vcpkg/binarycaching.cpp:2192]: Note to self/others: it would be really nice to not have yet another parser for this vs the one in DateVersion. No change requested.

File: include/vcpkg/binarycaching.private.h [hunk 9]
```diff
@@ -9,7 +9,14 @@
...
 {
-    std::string reformat_version(const std::string& version, const std::string& abi_tag);
+    // Turns:
+    // - <XXXX>-<YY>-<ZZ><whatever> -> <X>.<Y>.<Z>-vcpkg<abitag>
+    // - v?<X> -> <X>.0.0-vcpkg<abitag>
+    //   - this avoids turning 20-01-01 into 20.0.0-vcpkg<abitag>
+    // - v?<X>.<Y><whatever> -> <X>.<Y>.0-vcpkg<abitag>
+    // - v?<X>.<Y>.<Z><whatever> -> <X>.<Y>.<Z>-vcpkg<abitag>
+    // - anything else -> 0.0.0-vcpkg<abitag>
+    std::string reformat_version(StringView version, StringView abi_tag);
```
Comment (BillyONeal) [inline include/vcpkg/binarycaching.private.h:19]:
```suggestion
    std::string add_abitag_version_suffix(StringView version, StringView abi_tag);
```

**Review by BillyONeal — APPROVED** [review]
Since it doesn't look like vcpkg::reformat_version is exercised unless a nuget binary or asset cache is used, I would like to see some unit tests added for that (and renamed as commented before). Other than that LGTM!

**Comment by strega-nil-ms** [comment]
@BillyONeal there are unit tests for it already (hence why I'm confident that the new code is equivalent to the old code :P)

**Comment by BillyONeal** [comment]
> @BillyONeal there are unit tests for it already (hence why I'm confident that the new code is equivalent to the old code :P)

"cool beans"
